src/detect.c

```diff
@@ -2332,9 +2360,16 @@ static int SignatureCreateMask(Signature *s)
         SCLogDebug("sig requires http app state");
     }
 
-    if (s->sm_lists[DETECT_SM_LIST_HSBDMATCH] != NULL) {
-        s->mask |= SIG_MASK_REQUIRE_HTTP_STATE;
-        SCLogDebug("sig requires http app state");
+    if (s->sm_lists[DETECT_SM_LIST_FILEDATA] != NULL) {
+        if (s->alproto == ALPROTO_HTTP)
+            s->mask |= SIG_MASK_REQUIRE_HTTP_STATE;
+        else if (s->alproto == ALPROTO_SMTP)
+            s->mask |= SIG_MASK_REQUIRE_SMTP_STATE;
+        else if (s->alproto == ALPROTO_UNKNOWN) {
+            s->mask |= SIG_MASK_REQUIRE_HTTP_STATE;
```

Comment:
I think this is wrong. It sets a condition (rule requires http AND smtp state) that can't be true for a packet/flow. In such cases you can just not set the condition at all. (+ add comment)